File: 142.环形链表II/142.java
/**
 * 快慢指针解法
 */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head==null){ return null;}
        ListNode fast=head;
        ListNode slow=head;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
            if(fast==slow){
                //当快慢指针相遇
                ListNode temp=head;
                while(slow!=temp){
                    temp=temp.next;
                    slow=slow.next;
                }
                //当temp=slow时，说明在环的入口处相遇了，返回-
                //链表开始入环的第一个节点
                return temp;
            }
        }
        return null;
    }
}

File: 链表/142.环形链表II/142.js
/**
 * 快慢指针解法
 * 时间复杂度：O(N)，其中 N 是链表中的节点数
 * 空间复杂度：O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var detectCycle = function(head) {
    if(head==null){return null}
    let fast=head,slow=head
    while(fast && fast.next!==null){
        slow=slow.next;
        fast=fast.next.next;
        if(fast==slow){
            let temp=head;
            while(temp!==slow){
                temp=temp.next;
                slow=slow.next;
            }
            return temp
        }
    }
    return null
};